contracts/src/v0.8/dev/automation/2_1/AutomationRegistrar2_1.sol

```diff
@@ -0,0 +1,518 @@
+ *  calls `approve` to register upkeep and emit events to inform UI and others interested.
+ * Flow 2. auto approve ON / real time registration - UI calls `register` function as before, which calls the `registerUpkeep` function directly on
+ *  keeper registry and then emits approved event to finish the flow automatically without manual intervention.
+ * The idea is to have same interface(functions,events) for UI or anyone using this contract irrespective of auto approve being enabled or not.
+ * they can just listen to `RegistrationRequested` & `RegistrationApproved` events and know the status on registrations.
+ */
+contract AutomationRegistrar2_1 is TypeAndVersionInterface, ConfirmedOwner, ERC677ReceiverInterface {
+  /**
+   * DISABLED: No auto approvals, all new upkeeps should be approved manually.
+   * ENABLED_SENDER_ALLOWLIST: Auto approvals for allowed senders subject to max allowed. Manual for rest.
+   * ENABLED_ALL: Auto approvals for all new upkeeps subject to max allowed.
+   */
+  enum AutoApproveType {
+    DISABLED,
+    ENABLED_SENDER_ALLOWLIST,
+    ENABLED_ALL
+  }
+
+  bytes4 private constant REGISTER_REQUEST_SELECTOR = this.register.selector;
+
+  mapping(bytes32 => PendingRequest) private s_pendingRequests;
+  mapping(uint8 => UpkeepRegistrationState) private s_upkeepTypeConfigs;
+
+  LinkTokenInterface public immutable LINK;
+
+  /**
+   * @notice versions:
+   * - KeeperRegistrar 2.1.0: Update for compatability with registry 2.1.0
+   *                          Add auto approval levels by type
+   * - KeeperRegistrar 2.0.0: Remove source from register
+   *                          Breaks our example of "Register an Upkeep using your own deployed contract"
+   * - KeeperRegistrar 1.1.0: Add functionality for sender allowlist in auto approve
+   *                        : Remove rate limit and add max allowed for auto approve
+   * - KeeperRegistrar 1.0.0: initial release
+   */
+  string public constant override typeAndVersion = "AutomationRegistrar 2.1.0";
+
+  /**
+   * @notice UpkeepRegistrationState stores the auto-approval levels for upkeeps by type
+   * @member autoApproveType the auto approval setting (see enum)
+   * @member autoApproveMaxAllowed the max number of upkeeps that can be auto approved of this type
+   * @member approvedCount the count of upkeeps auto approved of this type
+   */
+  struct UpkeepRegistrationState {
+    AutoApproveType autoApproveType;
+    uint32 autoApproveMaxAllowed;
+    uint32 approvedCount;
+  }
+
+  /**
+   * @notice UpkeepTypeConfig configures the auto-approval levels for upkeeps by type
+   * @member autoApproveType the auto approval setting (see enum)
+   * @member autoApproveMaxAllowed the max number of upkeeps that can be auto approved of this type
+   */
+  struct UpkeepTypeConfig {
+    uint8 upkeepType;
+    AutoApproveType autoApproveType;
+    uint32 autoApproveMaxAllowed;
```

Comment:
these are also present in state. we only need them in one place?

Reply:
The `Config` struct is only used in the constructor and setter function. Similar to `config` and `state` structs in the registry.

Reply:
yeah but they don't have duplicated fields in registry. so technically the config fields shouldn't be part of state struct.

Reply:
I think it makes the constructor cleaner to use this config struct, ex...

```solidity
// with config struct
  constructor(
    address LINKAddress,
    address keeperRegistry,
    uint96 minLINKJuels,
    InitialTriggerConfig[] memory registrationConfigs
){ ... }
```

```solidity
// without
  constructor(
    address LINKAddress,
    address keeperRegistry,
    uint96 minLINKJuels,
    uint8[] triggerTypes;
    AutoApproveType[] autoApproveTypes,
    uint32[] autoApproveMaxAlloweds
){ ... }
```

the second feels messier to me, but I don't feel strongly

Reply:
I think of it like the state contains the config because part of the state is configurable, but part of the state is not

Reply:
agree the constructor is cleaner. I can see the case for having the two structs one just for input and one for storage but I think we just need to come up with better names. We use `struct Storage` in registry, maybe that would also make sense here?

Reply:
I renamed `InitialTriggerConfig` and `TriggerRegistrationState` also tried to make it clear in the natspec
```
// for InitialTriggerConfig
@dev this struct is only used in the constructor to set the initial values for various trigger configs
```